### PR TITLE
Added perlin noise and thresholding feature

### PR DESCRIPTION
**Pull request by pbaylies**
I've been adding a couple of features to my own personal branch, I've seen good results with the k_*_a (k-diffusion / ancestral) optimizers by using thresholding, aka limiting the range of the latents during optimization. Also added the ability in general to start with a percentage of perlin noise mixed into the initial noise, both features allow for more variations and options in the course of generating images.

Note: currently the thresholding feature is only implemented for k-diffusion style optimization. It also uses a fixed warmup schedule, which I have found helps.

### CODE REVIEW

**Comment by bakkot** [comment]
Neat! Would you mind posting some comparisons of outputs with these features enabled vs disabled? It would be helpful to see what effect they have.

(I can get to it at some point, but am not set up for it right now.)

**Comment by pbaylies** [comment]
Thanks @bakkot - simple thresholding example - prompt - "Stunning Portrait painting The Mona Lisa by Leonardo da Vinci" - compare thresholding at 0 (disabled) vs at 2.5:
image

**Comment by pbaylies** [comment]
Another example - adding 50% perlin noise:
image

**Comment by pbaylies** [comment]
One more - changing only the sampler from above:
image

**Comment by pbaylies** [comment]
Compare the above to thresholding turned off:
image

**Comment by lstein** [comment]
Thanks for your contributin. I'll need a day or two to wrap my head around this.

**Comment by pbaylies** [comment]
Thanks @lstein - my intuition about thresholding is that it helps stop overly large values from dominating the generation and therefore oversaturating / wiping out finer details; with it, I find that I can run at higher CFG values and with more steps and still get nice generations. The Mona Lisa example is a handy one because it oversaturates easily all by itself, though.

**Comment by pbaylies** [comment]
@lstein After further testing, it looks to me like issues with high CFG values are due to numerical overflow - setting a threshold of 1000 yields the same image as a threshold of a billion, but no threshold overflows. Lower thresholds are still useful for tweaking the image results as well.

Starting point:
FcChrnuXEAQrgFU


Compare vs. having a high threshold in place:

**Comment by pbaylies** [comment]
@lstein one more test / demonstration - no thresholding (3) vs 3 vs 2:

Settings:
FcCr3GQXEAEW6a0

No thresholding (0):


Threshold of 3:


Threshold of 2:

**Review by lstein — CHANGES_REQUESTED** [review]
I appreciate the work you're doing on this. I don't want to torture you with more change requests, but could you add a section in README.md that documents the perlin noise and thresholding feature? If it is too long to explain in a paragraph, then I'd encourage you to create a new perlin-specific README that explains what this feature is and what users can do with it. The examples you showed in the "portrait of a beautiful lady" comment would be a great start for this. (The mona lisa example, on the other hand...)

I'd like to see this feature go into the next release (after 1.14).

**Comment by pbaylies** [comment]
Thanks @lstein I've added a short description to the README along with a montage showing both the options in use.

File: docs/features/OTHER.md [hunk 1]
```diff
@@ -91,6 +91,20 @@ do not need to add up to 1.
...
 ---
 
+## Thresholding and Perlin Noise Initialization Options
+
+Two new options are the thresholding (`--threshold`) and the perlin noise initialization (`--perlin`) options. Thresholding limits the range of the latent values during optimization, which helps combat oversaturation with higher CFG scale values. Perlin noise initialization starts with a percentage (a value ranging from 0 to 1) of perlin noise mixed into the initial noise. Both features allow for more variations and options in the course of generating images.
+
+For better intuition into what these options do in practice, [here is a graphic demonstrating them both](static/truncation_comparison.jpg) in use. Perlin noise at initialization is varied going across by 0.0, 0.1, 0.2, 0.4, 0.5, 0.6,, 0.8, 0.9, 1.0; and the threshold is varied going down from 0, 1, 2, 3, 4, 5, 10, 20, 100. The other options are fixed, so the initial prompt is as follows:
+
+```
+    a portrait of a beautiful young lady -S 1950357039 -s 100 -C 20 -A k_euler_a --threshold 0 --perlin 0
```
Comment (lstein) [inline docs/features/OTHER.md:101]: Please add more examples of applying non-zero --perlin and --threshold options and reword the above paragraph. Right now, as it is written, the documentation makes it sound like these values are automatically scanned through, but I think you scanned through the range of values either by hand or by using a script.

Comment (pbaylies) [reply docs/features/OTHER.md:102]: Thanks! I've updated the PR.

**Comment by pbaylies** [comment]
Thanks @lstein I've updated the PR.

**Comment by lstein** [comment]
I was hoping to test this over the weekend, but my development environment is down due to a scheduled maintenance window. I'll aim to work on this Monday.

**Review by lstein — COMMENTED** [review]
I suspect that this isn't going to go in cleanly against development due to the time that has elapsed. I am checking now.

**Review by lstein — APPROVED** [review]

**Review by lstein — CHANGES_REQUESTED** [review]
As I was afraid, the WebGUI has moved on and the changes for perlin noise and thresholding do not make it into the UI. In addition, I notice that there is no support for these features in the dream.py CLI either, so I can't them out.

@pbaylies I know you've worked hard at this. Would you mind doing just one more update to make the feature available to the new WebGUI and the CLI?

Thanks!

**Comment by pbaylies** [comment]
Thanks @lstein see if that fixes things! In the Web UI, you should see my options at the bottom of the Sampler pane, not sure if it's still necessary to regenerate the frontend (if someone else had already, previously).
image

**Comment by pbaylies** [comment]
@lstein I've updated this again for the latest web rewrite; take a look?

**Review by lstein — CHANGES_REQUESTED** [review]
Almost there! A couple things I noticed this time through:
1. The perlin and threshold options are not getting logged in the dream log. The fix for this is here:
~~~~
diff --git a/ldm/dream/args.py b/ldm/dream/args.py
index 6cae4f6..a2ae896 100644
--- a/ldm/dream/args.py
+++ b/ldm/dream/args.py
@@ -185,6 +185,11 @@ class Args(object):
         switches.append(f'-W {a["width"]}')
         switches.append(f'-H {a["height"]}')
         switches.append(f'-C {a["cfg_scale"]}')
+        if a['perlin'] > 0:
+            switches.append(f'--perlin {a["perlin"]}')
+        if a['threshold'] > 0:
+            switches.append(f'--threshold {a["threshold"]}')
+            
         if a['grid']:
             switches.append('--grid')
         if a['seamless']:
~~~~
2. At certain values of perlin, I see this warning:
~~~~
/u/lstein/.conda/envs/ldm/lib/python3.9/site-packages/torch/functional.py:568: UserWarning: torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument. (Triggered internally at  /opt/conda/conda-bld/pytorch_1646756402876/work/aten/src/ATen/native/TensorShape.cpp:2228.)
  return _VF.meshgrid(tensors, **kwargs)  # type: ignore[attr-defined]
~~~~
3. You added new sampler types of `k_dpm_adaptive` and `k_dpm_fast` to the WebGUI, but these samplers are not defined in the kcrowson k_diffusion library. Selecting them causes the WebGUI to complain.
4. At step lengths of 50, the k_lms sampler is producing brown or washed out images with non-zero values of perlin noise. Is this expected? Other k samplers, including k_heun and k_euler_a, are working well, and I do see the expected extra variation.
5. The OTHER.md documentation file refers to an image file named `static/truncation_comparison.jpg`. Could you move it into docs/assets, where the other .md figures are located? `static` is for use by the web servers.

**Review by lstein — APPROVED** [review]

**Comment by lstein** [comment]
Thanks. I'll take another look!

**Pull request merged**